### PR TITLE
Datetime picker, splitter, item split, time-freeze

### LINKED ISSUE #2608: Time UI control to select a specific time instant
See the middle part of #2557 for the early discussion on this.

### PR DESCRIPTION
**Pull request by RacingTadpole**
Fixes #2608 and #2609 .

To do re the datetime picker:

- [x] I got this on "Total Vegetation Cover":
image
- [x] It's not obvious that you can click on the datetime to bring up the picker
- [x] Clicking off the picker should close it
- [x] Default to the first time in the list once you've picked a date
- [x] The width of the calendar icon and the date that replaces it are different, which makes the timeline change size when you pick a datetime.
- [x] The back arrow on the year display does nothing - maybe remove?
image
- [x] When a year or month has no dates, should it still highlight the row and display the hand pointer, even though clicking it does nothing? 
image
- [x] Slightly weird display here: 
image
- [ ] Test on Firefox, IE, Safari and Edge. (In particular is the backtick format going to work on IE?)
- [x] Now the picker can appear in the workbench, it can appear more than once. Currently when opened, the datepicker panel can appear under the wrong date.
- [x] Make the datepicker stateless, ie. when it is reopened, it uses the original time.
- [x] When the datepicker is closed, show the calendar icon, not the date. Put the date somewhere else.
- [x] Can you improve the look of the scroll bar here? 
image
- [ ] When you reach a point where there only a limited number of possible choices (~5-30?), just show them in a list. If there are more, there should be a screen for it - eg. for hours, minutes, seconds (or centuries etc). - Done this with year, but not a lower level yet, it's getting too complicated, i propose we leave it here for now. - Chloe 

To do re the split capability:
- [x] The latest versions of Chrome, Firefox, Internet Explorer, and Edge on desktop Chrome on Android.
- [x] Touch input on desktop and mobile in all the browsers above.
- [x] Currently using a unicode character for the face of the splitter thumb that is apparently not available on Android, so it looks like a box with an X through it. We could swap this for an SVG, or maybe use a different font? @chloeleichen?
- [x] Check Safari on desktop.
- [x] Check on iOS.
- [x] Check sharing (both per-layer split side and split position should be shared I'd say).
- [ ] Fix sharing mini-preview window to also clip layers (currently ignores the splitter) - 2D.
- [x] Fix sharing mini-preview window to also clip layers (currently ignores the splitter) - 3D.
- [x] Merge in Chloe's new date-time picker
- [ ] Check how stable item cloning really is when you click "split".
- [x] Add a checkbox to unfreeze the datetime.
- [ ] Improve the toggle-splitter icon (currently just a checkbox) @chloeleichen?
- [x] In 2D, moving the splitter triggers a feature info request again (@kring?)
- [x] When you open the datetime picker, then go back to the timeline, then when you go back to the datetime picker, the originally picked date is showing, not the latest one. Show the latest one.
- [x] When you turn on the splitter and there is a layer that is not a time based layer, we still show the timestamp selector in the workbench for that layer (this is the bug) together with the left/both/right (not a bug).
- [x] When you have the splitter active with an enabled layer in either the Left or Right configuration, using the new DateTimeControl in the workbench to select a new time shows the new imagery on Both sides. In order to get it back to just one side you have to press Both or the opposite side currently indicated, and then back to the side you want. (3D only.)
- [x] The `Split` action in the workbench control doesn't work, and just returns this error message: `No layer found
The WMS dataset "Total Vegetation Cover (PV + NPV) (copy)" has no layers matching "".` (I've seen it work before, maybe this is just something wrong with my build?)
- [x] When you resize the window (or do a window zoom in/out), the slider location and imagery gets out of sync.
- [x] The option to "split" does not re-appear after you have removed one of the duplicate items.
- [ ] Review.

Deferred:
- [ ] We could split region-mapped layers too.

(Some tasks courtesy of https://github.com/TerriaJS/terriajs/pull/2627.)

### CODE REVIEW

**Comment by philipgrimmett** [comment]
workbenchitem-date-timeline

datepicker-concept-1

**Comment by RacingTadpole** [comment]
@philipgrimmett - when you get a chance, could you design a "toggle splitter" button to replace my dumb checkbox? (Currently the checkbox shows whether the splitter is showing, not sure if you want two different icons to show that)
image

**Comment by RacingTadpole** [comment]
@kring - when you get a chance, I have a question about the right way to make "split" (duplicated) items sharable.  I've done [this](https://github.com/TerriaJS/terriajs/blob/2609_2608_all/lib/ReactViews/Workbench/Controls/ViewingControls.jsx#L73-L79), ie. after creating it (which automatically adds it to `terria.nowViewing`), I call `addUserCatalogMember(terria, newItem);` to add it to the user-added data group in the catalog.
If I don't do this, then "share" can't find it. But if I do it, then it serializes all the ImageryLayer `intervals` and `availableDimensions`, which are big. I can stop those being serialized, but I think there must be a better way, because: if I add a WMS url using "my data -> web data", then press share, it doesn't serialize the intervals or availableDimensions. 
How do I replicate that behaviour for my duplicated item?  Thanks!

**Comment by RacingTadpole** [comment]
@meh9, @kring - I'm ready to call the MVP version of this done, apart from including Phil's new toggle-splitter icons. 
Feel free to do a more thorough test of it any time.

**Comment by kring** [comment]
In a map that autoplays (such as TerriaMap):
* Add the `Chlorophyll-a concentration (annual)` layer or some other time-dynamic WMS.
* It will start playing immediately.  Press pause.
* Click the Timeline button on the workbench twice to toggle timeline sync'ing off and then back on.
* The timeline will reappear and be playing again.

I'm guessing we don't want this, right?  I think it would be better if we either remembered the previous "play" state, or if we always started paused after exiting non-timeline mode.  If this is hard I'm not too hung up on it, though.  In that case please change TerriaMap to not autoplay (just like everything else these days).

**Comment by kring** [comment]
A related problem that happens in every map, whether `autoPlay` is enabled or not:
* Add the Chlorophyll-a concentration (annual) layer or some other time-dynamic WMS.
* Select a new time by clicking on the timeline.  The date/time display on the workbench will update. Cool!
* Click the Timeline button on the workbench twice to toggle timeline sync'ing off and then back on.
* The current time will reset to the "default" (last) time instead of the one that was previously selected.

I think this one is a bit more serious than the above.  I definitely didn't expect it.  But maybe there's a good reason for it that I'm overlooking.

**Comment by RacingTadpole** [comment]
> The current time will reset to the "default" (last) time instead of the one that was previously selected.

Ironically if I've understood you right I worked quite hard to get this behaviour! :-)
If you choose a datetime with the chooser, then change to the timeline and pick a date using the timeline, the chooser shows the new date. I found it quite jarring if you then went to choose a new date with the chooser, and had it start you off at a different date to the one it was showing (eg, the one you last selected using the chooser).

**Comment by kring** [comment]
@RacingTadpole no that behavior sounds fine.  The problem is when switching back to "Timeline" mode.  I'd expect the timeline to change to reflect the currently selected date on the workbench.  Instead, when you click that Timeline button, the Timeline date and workbench date both change.

**Comment by kring** [comment]
The first click of the "Back" button in the date/time picker sometimes doesn't work.
* Add the Chlorophyll-a concentration (annual) layer or some other time-dynamic WMS.
* Use the date picker on the timeline to pick any time.
* Open the date picker again and click the back button.  The first click will be ignored, the second click will work.

I think the problem is that the logic in `goBack` in `DateTimePicker.jsx` isn't accounting for the drop into "select a time from a list" mode.

**Comment by kring** [comment]
Here's a case where where there's only one time in June 1999, but it makes me select a day of the month instead of dropping immediately into time picker mode.

select-day-instead-of-time

(used http://tds.gisclimatechange.ucar.edu/thredds/wms/products/products_best.ncd?service=WMS&version=1.3.0&request=GetCapabilities)

**Comment by kring** [comment]
These arrows:
image

Also when I click them, the list of days shifts around (as if I'm moving to a different month), but the label doesn't change.

**Comment by chloeleichen** [comment]
@kring fixed above issues

**Comment by meh9** [comment]
When initially adding a timeseries layer, it seems that the Imagery Time may be calculated using the wrong timezone. E.g. when first adding Solar Satellite, the Imagery Time in the workbench shows `01/01/2015, 10:00:00`, but the time on the timeline shows `01/01/2015, 11:00:00`.

Looking in a little more depth, I don't actually know where this 2015 is coming from, because 2015 is not covered by the data, and when you click back out to years we don't see a 2015. (However, this behaviour is consistent with pre-splitter behaviour - AREMI in production currently shows a 2015 date when you first add this layer.)

**Comment by kring** [comment]
Several problems shown in this gif:
no-time-picker
* When I click a date, it closes immediately even though there are a bunch of different times on that date.
* When I re-open the date picker, even though I clicked on the 13th, the 12th is highlighted and all the times listed are on the 12th.

I think what should happen is when I click on the 13th, it should show me the time picker with all the times on the 13th instead of using that dropdown below the calendar.  Also, if there are too many times on that one date, it should first go to an hour picker.

Testing with https://gamone.whoi.edu/ncWMS/wms?SERVICE=WMS&REQUEST=GetCapabilities&VERSION=1.3.0&DATASET=COAWST and then NECOFS_Boston -> salinity.

**Comment by kring** [comment]
Here's a good one to test with to make sure it drops into the hour picker correctly when selecting a day with heaps of times: http://mesonet.agron.iastate.edu/cgi-bin/wms/nexrad/n0r-t.cgi?service=WMS&version=1.3.0&request=GetCapabilities

**Comment by kring** [comment]
Some weirdness with the `Split` option in the workbench.  When I click Split, the Split option vanishes, but the new item that gets created has a Split option.  If I remove the item with the Split option, the Split option does not reappear on the other item.

File: lib/Models/CatalogItem.js [hunk 1]
```diff
@@ -566,6 +617,9 @@ CatalogItem.defaultPropertiesForSharing.push('isShown');
 CatalogItem.defaultPropertiesForSharing.push('isLegendVisible');
 CatalogItem.defaultPropertiesForSharing.push('nowViewingIndex');
 CatalogItem.defaultPropertiesForSharing.push('nowViewingCatalogItem');
+CatalogItem.defaultPropertiesForSharing.push('useOwnClock');
+CatalogItem.defaultPropertiesForSharing.push('canUseOwnClock');
```
Comment (kring) [inline lib/Models/CatalogItem.js]: Why share this?  It never changes, right?

**Comment by kring** [comment]
A problem introduced by me yesterday: all `TableCatalogItems` (CSV, etc.) now have the left/right/both controls when the splitter is active, even if they're not region mapped.  But non-region-mapped layers can't actually be split.

File: doc/connecting-to-data/catalog-items.md [hunk 2]
```diff
@@ -2,19 +2,18 @@ A Catalog Item is a dataset or service that can be enabled for display on the ma
 
 | Name | Type |
 |------|------|
-| [ABS ITT](catalog-type-details/abs-itt.md) | `abs-itt` |
```
[the remark this answers is not in the document]
Comment (RacingTadpole) [reply doc/connecting-to-data/catalog-items.md:5]: I'd like to, but only if we are sure no-one has any shared views from months ago that still have them. Can we check on Google Analytics?

**Comment by axman6** [comment]
I'm running into a bug with the time selector when using splits - it seems they're somehow linked, so changing the date is one side of the split also changes it in the other. In the video I've shared on slack, I'm trying to show the same layer at midday on the 30th of June and the 30th of December.

**Comment by kring** [comment]
@axman6 are you still seeing this problem:

> I'm running into a bug with the time selector when using splits - it seems they're somehow linked, so changing the date is one side of the split also changes it in the other. In the video I've shared on slack, I'm trying to show the same layer at midday on the 30th of June and the 30th of December.

**Comment by axman6** [comment]
@kring I believe this was fixed by @chloeleichen a while ago, and as far as my limited testing just now can tell, it appears to be all good - happy for this particular part of this issue to be closed.

**Comment by meh9** [comment]
@kring yes, that does make sense, and yes it is confusing. :)

If we were going to change something here, I think it should be that instead of defaulting to the last instant of a time period, we should default to the first instant. This way the time slider and the workbench would show the same time which would remove some confusion. I can't immediately see a downside to this (apart from the work involved to change it).

**Pull request merged**